Only restore a configured MAC addr on restart.

robmry opened the pull request:
**- What I did**

The API's `EndpointConfig` struct has a `MacAddress` field that's used for both the configured address, and the current address (which may be generated).

A configured address must be restored when a container is restarted, but a generated address must not.

The previous attempt to differentiate between the two, without adding a field to the API's `EndpointConfig` that would show up in 'inspect' output, was a field in the daemon's version of `EndpointSettings`, `MACOperational`. It did not work, `MACOperational` was set to 'true' when a configured address was used. So, it ensured addresses were regenerated, but failed to preserve a configured address.

Also, only fill in `Config.MacAddress` in 'inspect' output when the MAC address is configured.

Fixes https://github.com/moby/moby/issues/47228
Fixes https://github.com/moby/moby/issues/47146

**- How I did it**

So, this change removes code from https://github.com/moby/moby/pull/47168, and instead adds `DesiredMacAddress` to the wrapped version of EndpointSettings, which are persisted but do not form part of the 'inspect' output.

Its value is copied from MacAddress (the API field) when a container is created.

The second git commit fills in `Config.MacAddress` in inspect output using `DesiredMacAddress`, which is empty when the address is generated. It also ensures the field is filled in for a configured address on the default bridge network, before the container is started, by translating the network name `default` into `bridge`.

**- How to verify it**

New tests added.

On Windows - run a container like this ...

`docker run -ti --mac-address 00-11-22-33-44-55 mcr.microsoft.com/windows/nanoserver:ltsc2022 cmd.exe`

... and use `ipconfig /all` to check that the configured MAC address has been used. Also check `inspect` shows the correct address.

Stop the container, restart the daemon, and restart the container. Then, exec into the container, check that the configured MAC address is still in-use - and that the inspect output is still correct.

_This should be automated - https://github.com/moby/moby/issues/47283_

**- Description for the changelog**

Ensure that a generated MAC address is not restored when a container is restarted, but a configured MAC address is preserved. 

Containers created using 25.0.0 may have duplicate MAC addresses, they must be re-created.

Containers created using 25.0.0 or 25.0.1 with user-defined MAC addresses will get generated MAC addresses when they are started using 25.0.2. They must also be re-created.

robmry committed:
Only restore a configured MAC addr on restart.

The API's EndpointConfig struct has a MacAddress field that's used for
both the configured address, and the current address (which may be generated).

A configured address must be restored when a container is restarted, but a
generated address must not.

The previous attempt to differentiate between the two, without adding a field
to the API's EndpointConfig that would show up in 'inspect' output, was a
field in the daemon's version of EndpointSettings, MACOperational. It did
not work, MACOperational was set to true when a configured address was
used. So, while it ensured addresses were regenerated, it failed to preserve
a configured address.

So, this change removes that code, and adds DesiredMacAddress to the wrapped
version of EndpointSettings, where it is persisted but does not appear in
'inspect' results. Its value is copied from MacAddress (the API field) when
a container is created.

Signed-off-by: Rob Murray <rob.murray@docker.com>

akerouanton commented on api/types/network/endpoint.go:
```diff
@@ -14,7 +14,11 @@ type EndpointSettings struct {
+	// API param field) when the container is created, and cleared in the inspect
+	// output so that only MacAddress is returned to clients. DesiredMacAddress is
+	// used to set the value for netlabel.MacAddress passed to the driver.
+	DesiredMacAddress string `json:",omitempty"`
```
I'd even not expose it through the API, otherwise we'll have to support it until the API version v1.44 is dropped:

```suggestion
	DesiredMacAddress string `json:"-"`
```

robmry replied:
That'd stop it being encoded for persistent storage... so its value wouldn't be preserved over a daemon restart.

So, the idea is to deal with that by including it in the JSON when it has a value - but always clear it in the copy used for 'inspect' responses.

akerouanton replied:
Huh, right. I totally forgot we're using the same json tag for both HTTP and storage concerns.

corhere replied:
Why couldn't this field be added to `EndpointSettings`? 
```go
// EndpointSettings is a package local wrapper for
// networktypes.EndpointSettings which stores Endpoint state that
// needs to be persisted to disk but not exposed in the api.
```

robmry replied:
I'd forgotten / not-noticed that the comment says those extra fields are persisted - but I don't think they are?

I could be wrong but, I think, the wrapper is always reconstructed here when the container's started ...

https://github.com/moby/moby/blob/794f7127efd841e023dbd88051c577489f6cd3cf/daemon/container_operations.go#L758-L762

Also, the local-wrapper fields aren't available to the daemon's 'inspect' code - so it couldn't differentiate between a generated address and a configured one when it needed to fill in the container-wide `Config.MacAddress`.

(In the first PR, I got the logic wrong for deciding whether the address was configured - it wasn't as similar to the `operIPAM` case as we thought, so `operMAC` wasn't set up properly. And because the field isn't persisted, I don't think it was possible to tell the difference between a generated MAC and a configured one during the container-run, it has to be spotted and remembered during the create.)

If that seems right, I'll fix the comment.

corhere replied:
> I'd forgotten / not-noticed that the comment says those extra fields are persisted - but I don't think they are?

They're persisted in the container config: `(container.Container).NetworkSettings[].Networks[]`.

> I could be wrong but, I think, the wrapper is always reconstructed here when the container's started ...

So? That could be changed.

> Also, the local-wrapper fields aren't available to the daemon's 'inspect' code - so it couldn't differentiate between a generated address and a configured one when it needed to fill in the container-wide `Config.MacAddress`.

The local-wrapper fields are accessible from the `*Container`. I find it hard to believe that they are not available to the code which maps an ID to a `*Container` and then to an API response.

robmry replied:
Ok, thank you - that helped me to untangle it a bit.

akerouanton commented on api/types/network/endpoint.go:
```diff
@@ -24,6 +28,7 @@ type EndpointSettings struct {
 	IPv6Gateway         string
 	GlobalIPv6Address   string
 	GlobalIPv6PrefixLen int
+	MacAddress          string
```
No need to move that field, it's still what should be used by API clients to configure a container's MAC address.

akerouanton commented on daemon/inspect.go:
```diff
@@ -70,6 +70,7 @@ func (daemon *Daemon) ContainerInspectCurrent(ctx context.Context, name string,
 		if epConf.EndpointSettings != nil {
 			// We must make a copy of this pointer object otherwise it can race with other operations
 			apiNetworks[nwName] = epConf.EndpointSettings.Copy()
+			apiNetworks[nwName].DesiredMacAddress = ""
```
Ah, so you're not really sending it through the API since you blank it out here and it has `omitempty`. I think I still prefer to make the json pkg ignore that field altogether -- and then, no need to make it empty here.

robmry replied:
Yes, indeed - but, as noted above, it does need to be JSON encoded so that the configured value is stored. So, to keep it out of the 'inspect' output, it has to be cleared here.

robmry committed:
No inspect 'Config.MacAddress' unless configured.

Do not set 'Config.MacAddress' in inspect output unless the MAC address
is configured.

Also, make sure it is filled in for a configured address on the default
network before the container is started (by translating the network name
from 'default' to 'config' so that the address lookup works).

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry commented on daemon/container_operations.go:
```diff
@@ -752,11 +749,14 @@ func (daemon *Daemon) connectToNetwork(cfg *config.Config, container *container.
-		EndpointSettings: endpointConfig,
-		IPAMOperational:  operIPAM,
-		MACOperational:   operMAC,
+	eps := container.NetworkSettings.Networks[nwName]
+	if eps == nil {
+		eps = &network.EndpointSettings{
+			EndpointSettings: endpointConfig,
+		}
+		container.NetworkSettings.Networks[nwName] = eps
 	}
+	eps.IPAMOperational = operIPAM
```
I should have copied everything but the new field here - I'll change it.

corhere commented on integration/networking/mac_addr_test.go:
```diff
@@ -77,3 +79,154 @@ func TestMACAddrOnRestart(t *testing.T) {
+	const netName = "testcfgmacaddr"
+	network.CreateNoError(ctx, t, c, netName,
+		network.WithDriver("bridge"),
+		network.WithOption("com.docker.network.bridge.name", netName))
```
```suggestion
		network.WithOption(bridge.BridgeName, netName))
```
https://github.com/moby/moby/blob/10027e8c01ba98f2598f03bed177a71caec5b69c/libnetwork/drivers/bridge/labels.go#L5

robmry replied:
Done.

corhere reviewed (commented):
Everything not commented on LGTM

robmry commented on api/types/network/endpoint.go:
```diff
@@ -14,6 +14,9 @@ type EndpointSettings struct {
 	IPAMConfig *EndpointIPAMConfig
 	Links      []string
 	Aliases    []string // Aliases holds the list of extra, user-specified DNS names for this endpoint.
+	// MacAddress may contain a configured value from the API. Its value is copied
+	// into DesiredMacAddress when the container is created, after which MacAddress
+	// becomes operational data (it may contain a generated address).
```
Ok, I'll change it to ...

```
	// MacAddress may be used to specify a MAC address when the container is created.
	// Once the container is running, it becomes operational data (it may contain a
	// generated address).
```

robmry commented on daemon/container_operations.go:
```diff
@@ -692,10 +695,13 @@ func (daemon *Daemon) connectToNetwork(cfg *config.Config, container *container.
 	}
 	if endpointConfig == nil {
-		endpointConfig = &networktypes.EndpointSettings{}
+		endpointConfig = &network.EndpointSettings{}
+	}
+	if endpointConfig.EndpointSettings == nil {
+		endpointConfig.EndpointSettings = &networktypes.EndpointSettings{}
 	}
```
Done.

robmry commented on daemon/network/settings.go:
```diff
@@ -33,8 +33,9 @@ type Settings struct {
-	MACOperational bool
+	// DesiredMacAddress is the configured value, it's copied from MacAddress (the
+	// API param field) when the container is created.
+	DesiredMacAddress string
```
Yes, agreed - I dithered about it, but decided it was best to match the field it's related to.

robmry commented on integration/networking/mac_addr_test.go:
```diff
@@ -77,3 +79,154 @@ func TestMACAddrOnRestart(t *testing.T) {
+// Check that a configured MAC address is restored after a container restart,
+// and after a daemon restart.
+func TestCfgdMACAddrOnRestart(t *testing.T) {
+	skip.If(t, testEnv.DaemonInfo.OSType == "windows")
```
The issue with a MAC address generated by the bridge n/w driver getting reused after a container restart, then becoming a duplicate, doesn't apply on native-Windows - there's no bridge driver. (And these tests are all bridge-network based.)

Do we have any networking integration tests that start containers on Windows? There's probably a gap we should try to fill.

But, I think it has to be out of scope for this bug fix?

corhere replied:
The tests use the bridge driver, but are the daemon code paths being touched specific to bridge, or Linux? Because it looks suspiciously like the code is cross-platform.
 https://github.com/moby/moby/blob/cff4f20c44a3a7c882ed73934dec6a77246c6323/libnetwork/drivers/windows/windows.go#L539-L545

corhere replied:
> So, perhaps we can resolve this if I raise an issue saying the regression test is needed, and assign it to myself to have a go at?

I'd be satisfied with editing the "how to verify" section of the PR description to mention manually testing on Windows. Filing an issue to add a Windows regression test would be a huge bonus.

robmry replied:
Good stuff... that's done.

akerouanton reviewed (approved):
LGTM
